### PR TITLE
Bryanna/user profile page

### PR DESCRIPTION
**Pull request by bry-gavino**
Pages for Users to look at and edit their password and email.
(I added a "Users" link at the top of Layout just so it's easy to access it for now!)

## Screenshots

Screen Shot 2020-11-04 at 11 17 29 PM
Screen Shot 2020-11-04 at 11 17 36 PM

_Required for visual changes_

CC: @kwfk

### CODE REVIEW

File: .env.example [hunk 1]
```diff
@@ -1,4 +1,5 @@
-LOCAL_DATABASE_URL="postgresql://postgres@localhost:5432"
-DATABASE_URL="postgresql://postgres@localhost:5432/nbjc_development?schema=public"
+PG_USERNAME=postgres
+PG_PASSWORD=password
+PG_PORT=5432
```
Comment (kwfk) [inline .env.example]: this should match main

File: pages/users/sample.tsx [hunk 2]
```diff
@@ -0,0 +1,7 @@
+const SampleUser = {
+  ProfileType: 'Organization',
+  Email: 'oskibear@berkeley.edu',
+  Password: 'peepeepoo',
+};
+
+export default SampleUser;
```
Comment (kwfk) [inline pages/users/sample.tsx]: put this in `utils/sample_data.ts`

File: pages/users/settings_edit.tsx [hunk 3]
```diff
@@ -0,0 +1,22 @@
+import { FunctionComponent } from 'react';
+import Layout from 'components/Layout';
+import styles from 'styles/users/SettingsEdit.module.css';
+import SampleUser from './sample';
+import SettingsEdit from '../../components/user/edit';
+
+const UserProfSettings: FunctionComponent<{ initial?: number }> = () => {
```
Comment (kwfk) [inline pages/users/settings_edit.tsx]: use React.FunctionComponent or React.FC

File: styles/users/SettingsEdit.module.css [hunk 4]
```diff
@@ -0,0 +1,55 @@
+.content {
```
Comment (kwfk) [inline styles/users/SettingsEdit.module.css]: separate out the stylesheet between components and the page
(take a look at the other components to see how they're structured)

File: pages/users/settings.tsx [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+import { Button, Link } from '@material-ui/core';
+import CreateIcon from '@material-ui/icons/Create';
+import SettingsSave from 'components/user/save';
+import SettingsEdit from '../../components/user/edit';
+import SampleUser from './sample';
+
+const UserProfSettings: FunctionComponent<{ initial?: string }> = ({
+  initial = 'edit',
+}) => {
+  const [setting, setSetting] = useState(initial);
```
Comment (kwfk) [inline pages/users/settings.tsx]: initial doesn't need to be a prop
you can initialize state by putting in the `useState` like `useState('edit')`

File: pages/users/settings.tsx [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+  const hiddenPassword: string = '*'.repeat(SampleUser.Password.length);
```
Comment (kwfk) [inline pages/users/settings.tsx]: let's use a constant length like 6

File: pages/users/settings.tsx (hunk 5 above)
Comment (kwfk) [reply pages/users/settings.tsx]: it might be good to actually use integers instead of string

File: pages/users/settings.tsx (hunk 6 above)
Comment (kwfk) [reply pages/users/settings.tsx]: so just `hiddenPassword = '******'`

File: styles/users/SettingsEdit.module.css (hunk 4 above)
Comment (kwfk) [reply styles/users/SettingsEdit.module.css]: if the multiple components have the same styles you can import the same stylesheet to multiple components
feel free to ask me for clarification

File: pages/users/settings.tsx [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+              onClick={() => setSetting('edit')}
+              variant="contained"
+              color="primary"
+              disableElevation
+            >
+              Save
+            </Button>
+          </div>
+        </div>
+      </div>
```
Comment (kwfk) [inline pages/users/settings.tsx]: we can probably move the conditional inside the parent components

File: components/user/save.tsx [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+        defaultValue={email}
+        variant="outlined"
+        size="small"
+      />
+    </div>
+    <div className={styles.userInfo}>
+      <div className={styles.text}>Password</div>
+      <TextField
+        id="password"
+        defaultValue={password}
```
Comment (kwfk) [inline components/user/save.tsx]: don't show actual password

**Review by kwfk — CHANGES_REQUESTED** [review]
nice work! looks good but there are areas we can clean up
might need to talk to elizabeth about the editing flow
let's also add a cancel button

nice use of components :)

File: pages/users/settings.tsx [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+      <Button
+        variant="outlined"
+        color="primary"
+        disableElevation
+        startIcon={<CreateIcon />}
+        onClick={() => setSetting(1)}
+      >
+        Edit
+      </Button>
+    ) : undefined;
```
Comment (kwfk) [inline pages/users/settings.tsx]: let's use `null` instead of undefined

File: pages/users/settings.tsx [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+    ) : (
+      <SettingsSave
+        userType={SampleUser.ProfileType}
+        email={SampleUser.Email}
+        password={hiddenPassword}
+      />
+    );
+
+  const saveButton: typeof Button =
+    setting === 0 ? undefined : (
```
Comment (kwfk) [inline pages/users/settings.tsx]: null here too

File: pages/users/settings.tsx [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+import { Button, Link } from '@material-ui/core';
+import CreateIcon from '@material-ui/icons/Create';
+import SettingsSave from '../../components/user/Save/index';
+import SettingsEdit from '../../components/user/Edit/index';
+import { SampleUser } from '../../utils/sample-data';
+
+const UserProfSettings: React.FC = () => {
+  const [setting, setSetting] = useState(0);
+  const hiddenPassword = '******';
+  const editButton: typeof Button =
```
Comment (kwfk) [inline pages/users/settings.tsx]: you don't have to type the variable bc Typescript will implicitly know what it is

File: pages/users/settings.tsx [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+        color="primary"
+        disableElevation
+        startIcon={<CreateIcon />}
+        onClick={() => setSetting(1)}
+      >
+        Edit
+      </Button>
+    ) : undefined;
+
+  const userComponent: React.ReactElement =
```
Comment (kwfk) [inline pages/users/settings.tsx]: no type needed here too

File: pages/users/settings.tsx [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+      />
+    ) : (
+      <SettingsSave
+        userType={SampleUser.ProfileType}
+        email={SampleUser.Email}
+        password={hiddenPassword}
+      />
+    );
+
+  const saveButton: typeof Button =
```
Comment (kwfk) [inline pages/users/settings.tsx]: no type needed

File: components/user/Save/index.tsx [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+import styles from './Save.module.css';
+
+type Props = {
+  userType: string;
+  email: string;
+  password: string;
+};
+
+const SettingsSave: React.FunctionComponent<Props> = ({
+  userType,
```
Comment (kwfk) [inline components/user/Save/index.tsx]: remove unused prop

File: components/user/Edit/index.tsx [hunk 15]
```diff
@@ -0,0 +1,26 @@
+import styles from './Edit.module.css';
+
+type Props = {
+  userType: string;
+  email: string;
+  password: string;
+};
+
+const SettingsEdit: React.FunctionComponent<Props> = ({
+  userType,
```
Comment (kwfk) [inline components/user/Edit/index.tsx]: remove unused prop

File: styles/users/SettingsEdit.module.css (hunk 4 above)
Comment (kwfk) [inline styles/users/SettingsEdit.module.css]: rename file to `Settings.module.css`

File: components/user/Edit/Edit.module.css [hunk 16]
```diff
@@ -0,0 +1,14 @@
+.userInfo {
+  display: flex;
+  justify-content: space-between;
+  margin-top: 10px;
```
Comment (kwfk) [inline components/user/Edit/Edit.module.css]: add
```css
height: 40px;
align-items: center;
```

Comment (kwfk) [reply components/user/Edit/Edit.module.css]: this should be in the shared stylesheet

File: components/user/Edit/Edit.module.css [hunk 17]
```diff
@@ -0,0 +1,14 @@
...
+}
+
+.text {
```
Comment (kwfk) [inline components/user/Edit/Edit.module.css]: remove this style

File: styles/users/SettingsEdit.module.css [hunk 18]
```diff
@@ -0,0 +1,55 @@
...
+  height: 100%;
+  align-items: center;
+}
+
+.box {
+  width: 500px;
+  display: flex;
+  flex-direction: column;
+  justify-content: space-between;
+  margin-top: 10%;
```
Comment (kwfk) [inline styles/users/SettingsEdit.module.css]: change width to 60% and remove margin-top

File: styles/users/SettingsEdit.module.css [hunk 19]
```diff
@@ -0,0 +1,55 @@
+.content {
+  display: flex;
+  align-items: center;
+  flex-direction: column;
```
Comment (kwfk) [inline styles/users/SettingsEdit.module.css]: remove flex-direction

File: styles/users/SettingsEdit.module.css [hunk 20]
```diff
@@ -0,0 +1,55 @@
...
+  justify-content: center;
+  width: 100%;
+  height: 100%;
```
Comment (kwfk) [inline styles/users/SettingsEdit.module.css]: change height to `calc(100vh - 85px);`

**Review by kwfk — CHANGES_REQUESTED** [review]
I think the component naming is a little confusing. let's use `SettingsEdit` for when the user can edit the settings then `SettingsShow` as the viewing state

Also, move the component stylesheets to a shared stylesheet in `components/user/User.module.css` since it's exactly the same. Any comments added to the component stylesheet should also go in the same shared stylesheet

**Comment by bry-gavino** [comment]
Hey Fred!! I found that it was actually easier to not use components for the "change email" part. I also ended up making a new page for the "change password" part. As always,  let me know if you want me to change anything!
Screen Shot 2020-11-14 at 4 21 27 PM
Screen Shot 2020-11-14 at 4 21 47 PM
Screen Shot 2020-11-14 at 4 21 55 PM

File: components/user/EmailVerify/index.tsx [hunk 21]
```diff
@@ -0,0 +1,50 @@
+import { Link, Snackbar, Button } from '@material-ui/core';
+import styles from './EmailVerify.module.css';
+
+const EmailVerify: React.FC = () => {
+  const [open, setOpen] = React.useState(true);
```
Comment (kwfk) [inline components/user/EmailVerify/index.tsx]: import useState instead
```typescript
import { useState } from 'react'
```

File: pages/users/password_change.tsx [hunk 22]
```diff
@@ -0,0 +1,54 @@
+import Layout from 'components/Layout';
```
Comment (kwfk) [inline pages/users/password_change.tsx]: you want page components to use `-` instead of `_`

Comment (kwfk) [reply pages/users/password_change.tsx]: let's also rename the component to `ChangePassword`

File: pages/users/settings.tsx [hunk 23]
```diff
@@ -0,0 +1,95 @@
...
+import Layout from 'components/Layout';
+import styles from 'styles/users/Settings.module.css';
+import { Button, Link, TextField } from '@material-ui/core';
+import { sampleUserData } from '../../utils/sample-data';
+import ProgressStepper from '../../components/user/ProgressStepper/index';
+import EmailVerify from '../../components/user/EmailVerify/index';
+
+function capitalizeFirstLetter(string: string): string {
+  return string.charAt(0).toUpperCase() + string.slice(1);
+}
```
Comment (kwfk) [inline pages/users/settings.tsx]: we can use CSS `text-transform: capitalize` to capitalize text

File: components/user/ProgressStepper/index.tsx [hunk 24]
```diff
@@ -0,0 +1,43 @@
...
+
+function getSteps(): Array<string> {
+  return ['Not Started', 'In Progress', 'Submitted', 'Results'];
+}
+
+const ProgressStepper: React.FC = () => {
+  const [activeStep, setActiveStep] = React.useState(0);
+  const steps = getSteps();
+
+  const handleNext: void = () => {
```
Comment (kwfk) [inline components/user/ProgressStepper/index.tsx]: void should go after the parentheses

**Review by kwfk — COMMENTED** [review]
let's change the stepper to be highlighted at one step at a time
i.e. https://material-ui.com/components/steppers/#non-linear-alternative-label

File: pages/users/settings.tsx [hunk 25]
```diff
@@ -0,0 +1,95 @@
...
+            </Button>
+          </Link>
+        </div>
+      </div>
+    </div>
+  );
+
+  return (
+    <Layout title="User Profile Settings">
+      <EmailVerify />
```
Comment (kwfk) [inline pages/users/settings.tsx]: let's not have the email verify show for now (we may create a custom component for toasts later)

File: styles/users/Settings.module.css [hunk 26]
```diff
@@ -0,0 +1,79 @@
+.content {
+  display: flex;
+  align-items: center;
+  justify-content: center;
+  width: 100%;
+  height: calc(100vh - 85px);
```
Comment (kwfk) [inline styles/users/Settings.module.css]: change to `min-height`

File: pages/users/settings.tsx [hunk 27]
```diff
@@ -0,0 +1,95 @@
...
+    );
+
+  const passwordButton = (
+    <div className={styles.passwordButton}>
+      <div className={styles.emailEdit}>
+        <div>Password</div>
+        <div className={styles.emailButton}>
+          {hiddenPassword}
+          <Link href="../users/password_change">
+            <Button variant="outlined" color="primary" disableElevation>
```
Comment (kwfk) [inline pages/users/settings.tsx]: add onClick to go to ChangePassword page
see this to see how to use NextJS router to go to pages
https://github.com/calblueprint/nbjc/blob/a51a7af389736ee893665f6727ace20bf1a205b3/components/Layout.tsx#L46

File: pages/users/password_change.tsx [hunk 28]
```diff
@@ -0,0 +1,54 @@
...
+                variant="outlined"
+              />
+            </div>
+          </div>
+          <div className={styles.save}>
+            <Link href="../users/settings">
+              <Button color="primary" variant="contained" disableElevation>
+                Save Changes
+              </Button>
+            </Link>
```
Comment (kwfk) [inline pages/users/password_change.tsx]: see this to see how to use router to go to another page
https://github.com/calblueprint/nbjc/blob/a51a7af389736ee893665f6727ace20bf1a205b3/components/Layout.tsx#L46

File: pages/users/password_change.tsx [hunk 29]
```diff
@@ -0,0 +1,54 @@
...
+      <div className={styles.page}>
+        <div className={styles.content}>
+          <div className={styles.title}>Change Password</div>
+          <div className={styles.passwords}>
+            <div className={styles.field}>
+              Old Password
+              <TextField
+                size="small"
+                id="outlined-basic"
+                label={sampleUserData[0].email}
```
Comment (kwfk) [inline pages/users/password_change.tsx]: label is an email?

File: pages/users/password_change.tsx [hunk 30]
```diff
@@ -0,0 +1,54 @@
+import Layout from 'components/Layout';
+import styles from 'styles/users/PasswordChange.module.css';
+import { Button, TextField, Link } from '@material-ui/core';
+import { sampleUserData } from '../../utils/sample-data';
```
Comment (kwfk) [inline pages/users/password_change.tsx]: sample user data should not be used

File: components/user/ProgressStepper/index.tsx [hunk 31]
```diff
@@ -0,0 +1,43 @@
...
+import StepLabel from '@material-ui/core/StepLabel';
+import Button from '@material-ui/core/Button';
+import styles from './ProgressStepper.module.css';
+
+function getSteps(): Array<string> {
+  return ['Not Started', 'In Progress', 'Submitted', 'Results'];
+}
+
+const ProgressStepper: React.FC = () => {
+  const [activeStep, setActiveStep] = React.useState(0);
```
Comment (kwfk) [inline components/user/ProgressStepper/index.tsx]: import useState separately

File: styles/users/Settings.module.css [hunk 32]
```diff
@@ -0,0 +1,81 @@
+.content {
+  display: flex;
+  align-items: center;
+  justify-content: center;
+  width: 100%;
+  height: calc(100vh - 105px);
```
Comment (kwfk) [inline styles/users/Settings.module.css]: change to `min-height`

File: styles/users/PasswordChange.module.css [hunk 33]
```diff
@@ -0,0 +1,39 @@
...
+  align-items: center;
+  height: 60%;
+}
+
+.page {
+  display: flex;
+  flex-direction: column;
+  justify-content: center;
+  align-items: center;
+  height: calc(100vh - 105px);
```
Comment (kwfk) [inline styles/users/PasswordChange.module.css]: change to `min-height`

File: pages/users/settings.tsx [hunk 34]
```diff
@@ -0,0 +1,104 @@
...
+          <Link href="../users/password_change">
+            <Button
+              variant="outlined"
+              color="primary"
+              disableElevation
+              className={styles.fieldButton}
+            >
+              Edit
+            </Button>
+          </Link>
```
Comment (kwfk) [inline pages/users/settings.tsx]: use router to link
https://github.com/calblueprint/nbjc/blob/a51a7af389736ee893665f6727ace20bf1a205b3/components/Layout.tsx#L46

**Review by kwfk — CHANGES_REQUESTED** [review]
nice work! there some more minor changes needed here and there

## Overall Changes
center the ChangePassword component with the same CSS style as the ProfileSettings page

**Comment by kwfk** [comment]
I pushed some minor CSS refactoring so you'll want to pull before continuing

File: styles/users/PasswordChange.module.css [hunk 35]
```diff
@@ -0,0 +1,39 @@
...
+  align-items: center;
+  height: 60%;
+}
+
+.page {
+  display: flex;
+  flex-direction: column;
+  justify-content: center;
+  align-items: center;
+  min-height: calc(100vh - 85px);
```
Comment (kwfk) [inline styles/users/PasswordChange.module.css]: this should be `100vh - 105px`

File: styles/users/Settings.module.css [hunk 36]
```diff
@@ -0,0 +1,85 @@
+.content {
+  display: flex;
+  align-items: center;
+  justify-content: center;
+  width: 100%;
+  min-height: calc(100vh - 85px);
```
Comment (kwfk) [inline styles/users/Settings.module.css]: this should be `100vh - 105px`

File: pages/users/change-password.tsx [hunk 37]
```diff
@@ -0,0 +1,55 @@
...
+              <TextField
+                size="small"
+                id="confirm-password"
+                label="******"
+                variant="outlined"
+              />
+            </div>
+          </div>
+          <div className={styles.save}>
+              <Button color="primary" variant="contained" onClick={() => router.push('../users/settings')} disableElevation>
```
Comment (kwfk) [inline pages/users/change-password.tsx]: change to `/users/settings` no dots

File: pages/users/change-password.tsx [hunk 38]
```diff
@@ -0,0 +1,55 @@
+import Layout from 'components/Layout';
```
Comment (kwfk) [inline pages/users/change-password.tsx:1]: you'll need to run prettier here
you can do so with the shortcut `cmd+option+f`

File: pages/users/settings.tsx [hunk 39]
```diff
@@ -0,0 +1,97 @@
...
+  const passwordButton = (
+    <div className={styles.passwordButton}>
+      <div className={styles.field}>
+        <div>Password</div>
+        <div className={styles.fieldRight}>
+          {hiddenPassword}
+          <Button
+            variant="outlined"
+            color="primary"
+            onClick={() => router.push('../users/change-password')}
```
Comment (kwfk) [inline pages/users/settings.tsx]: change to `/users/change-password` no dots

**Review by kwfk — CHANGES_REQUESTED** [review]
Looking good! Mentioned that the stepper is currently overflowing. Other than that, it should be nearing approval!

File: components/user/ProgressStepper/ProgressStepper.module.css [hunk 40]
```diff
@@ -0,0 +1,54 @@
...
+  font-weight: 800;
+}
+
+.untouched {
+  color: rgb(143, 143, 143);
+}
+
+.progress {
+  width: 100%;
+  height: 15px;
```
Comment (kwfk) [inline components/user/ProgressStepper/ProgressStepper.module.css]: make height 12px to match designs

File: components/user/ProgressStepper/ProgressStepper.module.css [hunk 41]
```diff
@@ -0,0 +1,54 @@
...
+.current {
+  color: rgb(144, 86, 239);
+  font-weight: 800;
+}
+
+.untouched {
+  color: rgb(143, 143, 143);
+}
+
+.progress {
```
Comment (kwfk) [inline components/user/ProgressStepper/ProgressStepper.module.css:51]: add `border-radius: 36px`

**Review by kwfk — APPROVED** [review]
one small styling change then should be good!

**Pull request merged**